internal/controller/controllers/agent_controller.go

```diff
@@ -1761,3 +1767,99 @@ func (r *AgentReconciler) updateHostInstallProgress(ctx context.Context, host *m
 	})
 	return err
 }
+
+func (r *AgentReconciler) restoreHostByAgent(ctx context.Context, agent *aiv1beta1.Agent) error {
+	// Get InfraEnv
+	infraEnvName, exist := agent.Labels[aiv1beta1.InfraEnvNameLabel]
+	if !exist {
+		r.Log.Errorf("Failed to find infraEnv name for agent %s in namespace %s", agent.Name, agent.Namespace)
+		return errors.New("Missing InfraEnv name label")
+	}
+	key := types.NamespacedName{
+		Name:      infraEnvName,
+		Namespace: agent.Namespace,
+	}
+	infraEnv, err := r.Installer.GetInfraEnvByKubeKey(key)
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
+			r.Log.WithError(err).Errorf("InfraEnv with name '%s' is missing", infraEnvName)
+			return err
+		}
+		r.Log.WithError(err).Errorf("Failed to get InfraEnv with name '%s'", infraEnvName)
+		return err
+	}
+
+	// Get associated cluster if available
+	var clusterID *strfmt.UUID
+	var cluster *common.Cluster
+	if agent.Spec.ClusterDeploymentName != nil {
+		clusterKey := types.NamespacedName{
+			Namespace: agent.Spec.ClusterDeploymentName.Namespace,
+			Name:      agent.Spec.ClusterDeploymentName.Name,
+		}
+		cluster, err = r.Installer.GetClusterByKubeKey(clusterKey)
+		if err != nil {
+			r.Log.WithError(err).Errorf("Failed to get cluster (name: %s namespace: %s)",
+				agent.Spec.ClusterDeploymentName.Name, agent.Spec.ClusterDeploymentName.Namespace)
+			return err
+		}
+		clusterID = cluster.ID
+	}
+
+	host, err := createNewHost(agent, clusterID, infraEnv.ID)
+	if err != nil {
+		r.Log.WithError(err).Errorf("Failed to create Host with name '%s'", agent.Name)
+		return err
+	}
+
+	return r.Installer.CreateHostInKubeKeyNamespace(ctx, key, host)
+}
+
+func createNewHost(agent *v1beta1.Agent, clusterID, infraEnvID *strfmt.UUID) (*models.Host, error) {
+	if infraEnvID == nil {
+		return nil, errors.New("Missing InfraEnv ID")
+	}
+
+	// Create Intentory
+	hostInventory := models.Inventory{}
+	manufacturer, exist := agent.Labels[AgentLabelHostManufacturer]
+	if exist {
+		hostInventory.SystemVendor = &models.SystemVendor{}
+		hostInventory.SystemVendor.Manufacturer = manufacturer
+	}
+	inventory, err := json.Marshal(hostInventory)
+	if err != nil {
+		return nil, errors.New("Failed to marshal agent's inventory")
+	}
+
+	// Fetch State
+	var hostStatus string
+	if state, has_annotation := agent.GetAnnotations()[AgentStateAnnotation]; has_annotation {
+		hostStatus = state
+	} else {
+		if clusterID != nil {
+			hostStatus = models.HostStatusKnown
+		} else {
+			hostStatus = models.HostStatusKnownUnbound
+		}
+	}
+
+	// Create Host model
+	hostID := strfmt.UUID(agent.Name)
+	url := installer.V2GetHostURL{InfraEnvID: *infraEnvID, HostID: hostID}
+	host := &models.Host{
+		ID:            &hostID,
+		Href:          swag.String(url.String()),
```

Comment:
Is this set for the kubeAPI? I suppose it isn't expose to the user in any case, but I feel like this API wouldn't be available.